src/options.rs

```diff
@@ -0,0 +1,138 @@
+use crate::config::ConfigFile;
+use crate::error::Error;
+use crate::repo::{BlockStore, DataStore};
+use async_std::path::PathBuf;
+use bitswap::Strategy;
+use libp2p::core::{Multiaddr, PeerId};
+use libp2p::identity::Keypair;
+use std::borrow::Borrow;
+use std::fmt;
+
+/// All types can be changed at compile time by implementing
+/// `IpfsTypes`.
+pub trait IpfsTypes: Clone + 'static {
+    type TBlockStore: BlockStore;
+    type TDataStore: DataStore;
+    type TStrategy: Strategy;
+}
+
+/// Default IPFS types.
+#[derive(Clone, Debug)]
+pub struct Types;
+impl IpfsTypes for Types {
+    type TBlockStore = crate::repo::fs::FsBlockStore;
+    #[cfg(feature = "rocksdb")]
+    type TDataStore = crate::repo::fs::RocksDataStore;
+    #[cfg(not(feature = "rocksdb"))]
+    type TDataStore = crate::repo::mem::MemDataStore;
+
+    type TStrategy = bitswap::AltruisticStrategy;
+}
+
+/// Testing IPFS types
+#[derive(Clone, Debug)]
+pub struct TestTypes;
+impl IpfsTypes for TestTypes {
+    type TBlockStore = crate::repo::mem::MemBlockStore;
+    type TDataStore = crate::repo::mem::MemDataStore;
+    type TStrategy = bitswap::AltruisticStrategy;
+}
+
+/// Ipfs options
+#[derive(Clone)]
+pub struct IpfsOptions {
+    /// The path of the ipfs repo.
+    pub ipfs_path: PathBuf,
+    /// The keypair used with libp2p.
+    pub keypair: Keypair,
+    /// Nodes dialed during startup.
+    pub bootstrap: Vec<(Multiaddr, PeerId)>,
+    /// Enables mdns for peer discovery when true.
+    pub mdns: bool,
+}
+
```

Comment:
The `fn inmemory_generated_keys(mdns: bool) -> IpfsOptions<TestTypes>` is missing. I am fine with other names but there needs to be a way to create an IpfsOptions for tests easily, without any IO. I'm not sure if the `Default` impl should do that, maybe not change the semantics at least yet?